docs/tutorials/quantum_reinforcement_learning.ipynb

```diff
@@ -0,0 +1,1564 @@
+      "Episode 770/2000, average last 10 rewards 443.9\n",
+      "Episode 780/2000, average last 10 rewards 494.2\n",
+      "Episode 790/2000, average last 10 rewards 333.1\n",
+      "Episode 800/2000, average last 10 rewards 367.1\n",
+      "Episode 810/2000, average last 10 rewards 317.8\n",
+      "Episode 820/2000, average last 10 rewards 396.6\n",
+      "Episode 830/2000, average last 10 rewards 494.1\n",
+      "Episode 840/2000, average last 10 rewards 500.0\n"
+     ]
+    }
+   ],
+   "source": [
+    "env = gym.make(\"CartPole-v1\")\n",
+    "\n",
+    "interaction = namedtuple('interaction', ('state', 'action', 'reward', 'next_state', 'done'))\n",
+    "    \n",
+    "episode_reward_history = []\n",
+    "step_count = 0\n",
+    "for episode in range(n_episodes):\n",
+    "    episode_reward = 0\n",
+    "    state = env.reset()\n",
+    "    done = False\n",
+    "    \n",
+    "    while not done:\n",
+    "        # Increase step count\n",
+    "        step_count += 1\n",
+    "\n",
+    "        # Interact with env\n",
+    "        state, action, next_state, reward, done = interaction_env(state, model, epsilon, n_actions, env)\n",
+    "        \n",
+    "        # Store last interaction in the replay memory\n",
+    "        sarsd = interaction(np.copy(state), action, reward, np.copy(next_state), float(done))\n",
+    "        replay_memory.append(sarsd)\n",
+    "        \n",
+    "        state = np.array(next_state)\n",
+    "        episode_reward += reward\n",
+    "        \n",
+    "        # Update model\n",
+    "        if step_count % update_VQC == 0 and len(replay_memory) >= batch_size:\n",
+    "            # Sample a batch of interactions\n",
+    "            batch = random.choices(replay_memory, k=batch_size)\n",
+    "            batch = interaction(*zip(*batch))\n",
+    "            \n",
+    "            # Update Q-function\n",
+    "            Q_learning_update(np.asarray(batch.state),\n",
+    "                              np.asarray(batch.reward, dtype=np.float32),\n",
+    "                              np.asarray(batch.next_state),\n",
+    "                              np.asarray(batch.action),\n",
+    "                              np.asarray(batch.done, dtype=np.float32),\n",
+    "                              model, gamma, n_actions)\n",
+    "        \n",
+    "        # Update target model\n",
+    "        if episode % update_target == 0:\n",
+    "            model_target.set_weights(model.get_weights())\n",
+    "\n",
+    "    # Decay epsilon\n",
+    "    epsilon = max(epsilon * decay_epsilon, epsilon_min)\n",
+    "    episode_reward_history.append(episode_reward)\n",
+    "    if (episode+1)%10 == 0:\n",
+    "        avg_rewards = np.mean(episode_reward_history[-10:])\n",
+    "        print(\"Episode {}/{}, average last 10 rewards {}\".format(\n",
+    "            episode+1, n_episodes, avg_rewards))\n",
+    "        if avg_rewards >= 500.0:\n",
+    "            break"
```

Comment:
If we are going to have a second environment interaction function there is no need for this training loop function to remain this complicated. In particular in the last review I proposed this change:

```python3
env = gym.make("CartPole-v1")
env.reset()

# Gather initial batch_size replays.
replay_memory += gather_steps(state_bounds, n_actions, model, batch_size, env, False, 1.0)

for episode in range(n_episodes):
    episode_reward = 0
    env.reset()
    
    while True:
        # Gather and store n_steps_update new episode steps.
        trajectories = gather_steps(
            state_bounds, n_actions, model, n_steps_update, env, False, epsilon)
        replay_memory += trajectories

        # Sample batch_size steps and train.
        training_batch = np.random.choice(replay_memory, size=batch_size)
        q_learning_update(
          np.asarray([x['state'] for x in training_batch]),
          np.asarray([x['reward'] for x in training_batch], dtype=np.float32),
          np.asarray([x['next_state'] for x in training_batch]),
          np.asarray([x['action'] for x in training_batch]),
          np.asarray([x['done'] for x in training_batch], dtype=np.float32),
          gamma, n_actions)

        episode_reward += np.sum([x['reward'] for x in trajectories])
    
        # Update target model. Is this correct ?
        if episode % update_target == 0:
            model_target.set_weights(model.get_weights())

        # The episode is finished.
        if trajectories[-1]['done']:
            break

    # Decay epsilon.
    epsilon = max(epsilon * decay_epsilon, epsilon_min)
    episode_reward_history.append(episode_reward)
    if (episode+1)%10 == 0:
        avg_reward = np.mean(episode_reward_history[-10:])
        print("Episode {}/{}, average last 10 rewards {}".format(
            episode+1, n_episodes, avg_reward))
        if avg_reward >= 500.0:
            break
```
Which entirely eliminates the need for the `if step_count % update_VQC ==0 and len(...) >= memory_size`, `done` and the variable `state` from the training loop, as well as the need for a namedtuple. In the snippet I use `gather_step`, but this can be replaced with `interaction_env` and a small change to interaction_env to add a variable number `k` steps from the environment at a time. I feel that this is a very useful simplification to this training loop and is something worth incorporating for ease of understanding for the readers.